File: group_7/src/kernel/src/gdt.c
#include <stdint.h>
#include "gdt.h"
// TODO: Get the GDT entries from somewhere?

/*
------------------------------------------------------------------------------
This file contains the implementation of the global description table.
------------------------------------------------------------------------------
*/
#define GDT_ENTRIES 5

struct gdt_entry gdt[GDT_ENTRIES];
struct gdt_ptr gdt_ptr;

void gdt_set_gate(int32_t num, uint32_t base, uint32_t limit, uint8_t access, uint8_t gran)
{
    gdt[num].base_low = (base & 0xFFFF);
    gdt[num].base_middle = (base >> 16) & 0xFF;
    gdt[num].base_high = (base >> 24) & 0xFF;

    gdt[num].limit_low = (limit & 0xFFFF);
    gdt[num].granularity = (limit >> 16) & 0x0F;

    gdt[num].granularity |= gran & 0xF0;
    gdt[num].access = access;
}

void gdt_load(struct gdt_ptr *gdt_ptr)
{
    asm volatile("lgdt %0"
                 :
                 : "m"(*gdt_ptr));
}

void init_gdt()
{
    // Set the GDT limit
    gdt_ptr.limit = sizeof(struct gdt_entry) * GDT_ENTRIES - 1;
    gdt_ptr.base = (uint32_t)&gdt;

    // num, base, limit, access, granularity
    gdt_set_gate(0, 0, 0, 0, 0);                // Null segment
    gdt_set_gate(1, 0, 0xFFFFFFFF, 0x9A, 0xCF); // Code segment
    gdt_set_gate(2, 0, 0xFFFFFFFF, 0x92, 0xCF); // Data segment
    gdt_set_gate(3, 0, 0xFFFFFFFF, 0xFA, 0xCF); // User mode code segment
    gdt_set_gate(4, 0, 0xFFFFFFFF, 0xF2, 0xCF); // User mode data segment

    // Load the GDT
    gdt_load(&gdt_ptr);

    // Flush GDT pointer
    // TODO Implement this?
    // gdt_flush((uint32_t)&gdt_ptr);
}

File: group_OSDev_18/src/gdt.zig
const std = @import("std");

const GdtEntry = packed struct {
    limit_low: u16,
    base_low: u16,
    base_middle: u8,
    access: u8,
    granularity: u8,
    base_high: u8,

    fn init(base: u32, limit: u32, access: u8, granularity: u8) GdtEntry {
        return .{
            .base_low = @truncate(u16, base & 0xFFFF),
            .base_middle = @truncate(u8, (base >> 16) & 0xFF),
            .base_high = @truncate(u8, (base >> 24) & 0xFF),
            .limit_low = @truncate(u16, limit & 0xFFFF),
            .granularity = @truncate(u8, ((limit >> 16) & 0x0F) | (granularity & 0xF0)),
            .access = access,
        };
    }
};

const GdtRegister = packed struct {
    limit: u16,
    base: *[5]GdtEntry,

    fn init(table: *[5]GdtEntry) GdtRegister {
        return .{
            .limit = @as(u16, @sizeOf(@TypeOf(table.*))) - 1,
            .base = table,
        };
    }
};

extern fn loadGdt(register: *const GdtRegister) void;

var gdt_table: [5]GdtEntry = undefined;
var gdt_register: GdtRegister = undefined;

pub fn init() void {
    gdt_table = [5]GdtEntry{
        GdtEntry.init(0, 0x00000000, 0x00, 0x00), // Null Descriptor
        GdtEntry.init(0, 0xFFFFFFFF, 0x9A, 0xCF), // Kernel Mode Code Segment
        GdtEntry.init(0, 0xFFFFFFFF, 0x92, 0xCF), // Kernel Mode Data Segment
        GdtEntry.init(0, 0xFFFFFFFF, 0xFA, 0xCF), // User Mode Code Segment
        GdtEntry.init(0, 0xFFFFFFFF, 0xF2, 0xCF), // User Mode Data Segment
    };
    gdt_register = GdtRegister.init(&gdt_table);
    loadGdt(&gdt_register);
}
